omegaconf/omegaconf.py

```diff
@@ -585,15 +586,14 @@ def is_missing(cfg: Any, key: DictKeyType) -> bool:
         except (UnsupportedInterpolationType, KeyError, AttributeError):
             return False
 
+    # DEPRECATED: remove in 2.2
     @staticmethod
     def is_optional(obj: Any, key: Optional[Union[int, str]] = None) -> bool:
-        if key is not None:
-            assert isinstance(obj, Container)
-            obj = obj._get_node(key)
-        if isinstance(obj, Node):
-            return obj._is_optional()
-        else:
-            return True
+        warnings.warn(
+            "`OmegaConf.is_optional()` is deprecated, see https://github.com/omry/omegaconf/issues/698",
+            stacklevel=2,
+        )
+        return _is_optional(obj, key)
```

Comment:
I am not sure there is a good reason behind defaulting to True.
If there isn't any, I think you can maintain the old deprecated implementation here for now and make a more strict internal _is_optional.

Reply:
I think that the defaulting to True is for the case where `_is_optional(obj, key)` is called and `key` cannot be found.

Currently, no code in the `omegaconf/` directory is using this "defaulting to True" branch, as all the calls use the one-argument signature `_is_optional(obj)` (though some code in the `tests/` directory does use the two-argument signature).

Instead of returning `True` in the case where `key` cannot be found, what do you think about raising `ConfigKeyError`?

Reply:
Since it's an internal function and the branch is not used now, I suggest using assertion to ensure that it's not tripping anyone.

Reply:
> Make a pass on the usage to convince yourself it's not current;y possible for a user to hit this branch though.

Yup, it's not possible for a user to hit.